tensorflow_addons/utils/test_utils.py

```diff
@@ -80,17 +80,79 @@ def create_virtual_devices(
-            tf.config.experimental.VirtualDeviceConfiguration(
-                memory_limit=memory_limit_per_device
-            )
+            tf.config.LogicalDeviceConfiguration(memory_limit=memory_limit_per_device)
             for _ in range(num_devices)
         ],
     )
 
-    return tf.config.experimental.list_logical_devices(device_type)
+    return tf.config.list_logical_devices(device_type)
+
+
+def create_or_get_logical_devices(
+    num_devices, force_device=None, memory_limit_per_device=1024
+):
+    """Virtualize a the physical device into logical devices or get devices if virtualization
+    has already occurred.
+
+    Args:
+        num_devices: The number of virtual devices needed.
+        force_device: 'CPU'/'GPU'. Defaults to None, where the
+            devices is selected based on the system.
+        memory_limit_per_device: Specify memory for each
+            virtual GPU. Only for GPUs.
+
+    Returns:
+        logical_devices_out: A list of logical devices which can be passed to
+            tf.distribute.MirroredStrategy()
+    """
+    if force_device is None:
+        device_type = (
+            "GPU" if len(tf.config.list_physical_devices("GPU")) > 0 else "CPU"
+        )
+    else:
+        assert force_device in ["CPU", "GPU"]
+        device_type = force_device
+
+    physical_devices = tf.config.list_physical_devices(device_type)
+
+    # check the logical device configuration. Do not use list device because that actually initializes devices.
+    logical_config = tf.config.get_logical_device_configuration(physical_devices[0])
+
+    # explicitly confirm that we have no device configuration.
+    if logical_config is None:
+        # create devices
+        logical_devices = create_virtual_devices(
+            num_devices, force_device, memory_limit_per_device
+        )
+        logging.info("%i logical devices initialized" % num_devices)
+    else:
+        # if we have a configuration, then get the logical devices.
+        logical_devices = tf.config.list_logical_devices(device_type)
+
+    # take at most num_devices number of logical devices.
+    logical_devices_out = logical_devices[:num_devices]
+
+    # confirm that we are returning the correct number of logical devices.
+    if len(logical_devices_out) < num_devices:
+        raise RuntimeError(
+            """%i logical devices have been initialized at an earlier stage,
+        but the current request is for %i logical devices. Please initialize more logical devices at the earlier stage.
+        You are seeing this error because you cannot modify logical devices after initialization.
+        """
+            % (len(logical_devices), num_devices)
+        )
+    elif len(logical_devices_out) > num_devices:
+        raise ArithmeticError(
+            """List comprehension failure.
+        Expected the following python code to return true, when x, i, n are integers:
+        x <= len([i for i in range(n)][:x])
+        """
+        )
```

Comment:
Use an `assert` instead of a conditional. More importantly, this is a useless check, since we can assume that python isn't broken.

Reply:
makes sense I'll remove that check and chance the others to asserts